### ISSUE TITLE
Samsung Android 4.2.2 doesn't work correctly

### ISSUE THREAD

**Issue by marcucio** [mozilla/localForage#358]
On Samsung 4.2.2 device, localforage defaults to websql, when I try to add something I get the following error:

TypeError: Cannot call method 'transaction' of undefined

I'm looking into it now to see exactly where it fails but this is the code that will make it fail, this works find on other browser/apps I tested on:

```
	var data = {
		id: 12345,
		name: 'hello world'
	};
	var key = 'testData';

		localforage.config({
			name        : 'ManageIt',
			version     : 1.0,
			storeName   : 'data',
			driver: [localforage.WEBSQL]
		});

	localforage.setItem(key, data, function (err) {
		if(err) {
			//I GET THE ERROR HERE: TypeError: Cannot call method 'transaction' of undefined
			mig.utils.error('ObjectStore.prototype.setItem key:'+key+ ' err:'+err +' data:'+data);
		}
		self.getData('testData', function (dataReturned) {
			callback();
		});
	});
```

**Comment by marcucio** [comment]
Just a little more info:

```
dbInfo.db = openDatabase(dbInfo.name, String(dbInfo.version), dbInfo.description, dbInfo.size);
```

if failing with {"message":"SECURITY_ERR: DOM Exception 18","name":"SECURITY_ERR","code":18} and localforage is properly setting the driver to use localstorage:
```
return self.setDriver(self.LOCALSTORAGE).then(function() {
    return self._initStorage(options);
}).then(resolve)["catch"](reject);
```

but it looks like we still try to use setItem for the the sql driver thus dbInfo.db is not defined. I'm guessing the proper functionality should be that after the fall over to localstorage we should use localstorage not websql. It works fine for the second call we make to localforage so we do fall over properly for all additional calls.

I guess my question is should we be able to recover the first call before we fall over to localstorage?

**Comment by marcucio** [comment]
last comment for anyone who runs into this, you can work around this issue by calling localforage.ready() before you make any localforage set/get calls. I don't think this is how the driver is supposed to work but it is a workaround if you need it to work on older android devices.

**Comment by tofumatt** [comment]
Yeah, this should definitely be happening automatically, I’ll try to look into this although I don’t have any Samsung devices myself.

Thanks!

**Comment by marcucio** [comment]
you can test this functionality by just making "dbInfo.db = openDatabase(dbInfo.name, String(dbInfo.version),  dbInfo.description, dbInfo.size);" fail, maybe just throw an error instead of making the openDatabase call, that should simulate a openDatabase fail

**Comment by johannds** [comment]
Is this still an issue? I'm looking to use this on a deployment to about 30 000 devices, most of which are running Android.

**Comment by tofumatt** [comment]
I still haven't had the chance to test further as I don't have any Samsung devices and have been looking at other issues, but we have recently landed some Android fixes. It's worth testing again on one device and reporting back.

**Comment by happybridge14** [comment]
@marcucio Are you using the dist file? I saw the same error information. When I was using localforage.WEBSQL only, the error info is "Failed to execute 'transaction' on 'Database': The callback provided as parameter 1 is not a function.". If I use localforage.LOCALSTORAGE only, the error info is "Cannot read property 'transaction' of undefined". The files in the "src" folder are working well, but the files in the "dist" folder are not working.

**Comment by cristoferdomingues** [comment]
Hello Guys, has this issue resolved?

**Comment by tofumatt** [comment]
Again, I'm not 100% sure. We did land fixes that I would think cleared it up, but haven't had anyone confirm.

**Comment by marcucio** [comment]
should be fixed already with #511 I can see if I can dig up this old device again to see if it works but I'm pretty sure it is the same issue as #511 

We can probably close this issue out

**Comment by tofumatt** [comment]
Okay then, closed for now, but feel free to reopen if we closed this too quick! :smile:
